src/audio/mfcc/mfcc_init.c

```diff
@@ -0,0 +1,453 @@
+ * Also this is a too long function. Split to STFT, Mel filter, etc. parts.
+ */
+int mfcc_setup(struct comp_dev *dev, int max_frames, int sample_rate, int channels)
+{
+	struct mfcc_comp_data *cd = comp_get_drvdata(dev);
+	struct sof_mfcc_config *config = cd->config;
+	struct mfcc_state *state = &cd->state;
+	struct mfcc_fft *fft = &state->fft;
+	struct psy_mel_filterbank *fb = &state->melfb;
+	struct mfcc_dct *dct = &state->dct;
+	int ret;
+
+#ifdef MFCC_DEBUGFILES
+	mfcc_init_debug_open();
+	mfcc_generic_debug_open();
+#endif
+
+	comp_dbg(dev, "mfcc_setup()");
+
+	/* Check currently hard-coded features to match configuration request */
+	if (!config->round_to_power_of_two || !config->snip_edges ||
+	    config->subtract_mean || config->use_energy) {
+		comp_err(dev, "mfcc_setup(): Can't change currently hard-coded features");
+		return -EINVAL;
+	}
+
+	if (config->sample_frequency != sample_rate) {
+		comp_err(dev, "mfcc_setup(): Config sample_frequency does not match stream");
+		return -EINVAL;
+	}
+
+	cd->max_frames = max_frames;
+	state->sample_rate = sample_rate;
+	state->low_freq = config->low_freq;
+	state->high_freq = (config->high_freq == 0) ? (sample_rate >> 1) : config->high_freq;
+	if (state->low_freq > state->high_freq) {
+		comp_err(dev, "mfcc_setup(): Config high_freq must be larger than low_freq");
+		return -EINVAL;
+	}
+
+	comp_info(dev, "mfcc_setup(), source_channel = %d, stream_channels = %d",
+		  config->channel, channels);
+	if (config->channel >= channels) {
+		comp_err(dev, "mfcc_setup(): Illegal channel");
+		return -EINVAL;
+	}
+
+	if (config->channel < 0)
+		state->source_channel = 0;
+	else
+		state->source_channel = config->channel;
+
+	state->emph.enable = config->preemphasis_coefficient > 0;
+	state->emph.coef = -config->preemphasis_coefficient; /* Negate config parameter */
+	fft->fft_size = config->frame_length;
+	fft->fft_padded_size = 1 << (31 - norm_int32(fft->fft_size)); /* Round up to nearest 2^N */
+	fft->fft_hop_size = config->frame_shift;
+	fft->half_fft_size = (fft->fft_padded_size >> 1) + 1;
+
+	comp_info(dev, "mfcc_setup(), emphasis = %d", config->preemphasis_coefficient);
+	comp_info(dev, "mfcc_setup(), fft_size = %d", fft->fft_size);
+	comp_info(dev, "mfcc_setup(), fft_padded_size = %d", fft->fft_padded_size);
+	comp_info(dev, "mfcc_setup(), fft_hop_size = %d", fft->fft_hop_size);
+	comp_info(dev, "mfcc_setup(), fft->half_fft_size = %d", fft->half_fft_size);
+
+	/* Calculated parameters */
+	state->prev_data_size = fft->fft_size - fft->fft_hop_size;
+	state->buffer_size = fft->fft_size + max_frames;
+
+	/* Allocate buffer input samples and overlap buffer */
+	state->sample_buffers_size = sizeof(int16_t) *
+		(state->buffer_size + state->prev_data_size + fft->fft_size);
+
+	comp_info(dev, "mfcc_setup(), buffer_size = %d, prev_size = %d",
+		  state->buffer_size, state->prev_data_size);
+
+	state->buffers = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
```

Comment:
it's ok for the buffers to be cacheable?

Reply:
Yes, cache is needed for performance. These are not shared across cores.

Reply:
Also this data is private for every instance of MFCC if ever more than one. So cacheable (for performance) is OK, right?

Reply:
There are several things that make cached data access difficult, particularly on incoherent architectures like Xtensa: (1) cross-core access, (2) DMA, (3) mixed cached and uncached access, (4) manual cache manipulation: invalidation and writeback (have I forgotten anything?). If none of those is done here then it should be safe

Reply:
Thanks for good input (to some section in SOF docs?). I think none of 1-4 applies for this usage.